tool/tbot/main.go

```diff
@@ -0,0 +1,607 @@
+
+	conditions := role.GetImpersonateConditions(types.Allow)
+	return conditions.Roles, nil
+}
+
+// describeIdentity writes an informational message about the given identity to
+// the log.
+func describeIdentity(ident *identity.Identity) error {
+	tlsIdent, err := tlsca.FromSubject(ident.XCert.Subject, ident.XCert.NotAfter)
+	if err != nil {
+		return trace.WrapWithMessage(err, "bot TLS certificate can not be parsed as an identity")
+	}
+
+	// certType describes which type of certificates this is
+	var certType string
+	if tlsIdent.Impersonator == "" {
+		certType = "bot"
+	} else {
+		certType = "impersonated"
+	}
+
+	var principals []string
+	for _, principal := range ident.Cert.ValidPrincipals {
+		if !strings.HasPrefix(principal, nologinPrefix) {
+			principals = append(principals, principal)
+		}
+	}
+
+	duration := time.Second * time.Duration(ident.Cert.ValidBefore-ident.Cert.ValidAfter)
+	log.Infof(
+		"Successfully fetched new %s certificates, now valid: after=%v, before=%v, duration=%s, renewable=%v, disallow-reissue=%v, roles=%v, principals=%v",
+		certType,
+		time.Unix(int64(ident.Cert.ValidAfter), 0),
+		time.Unix(int64(ident.Cert.ValidBefore), 0),
+		duration,
+		tlsIdent.Renewable,
+		tlsIdent.DisallowReissue,
+		tlsIdent.Groups,
+		principals,
+	)
+
+	return nil
+}
+
+func renewLoop(cfg *config.BotConfig, client *auth.Client, ident *identity.Identity) error {
+	// TODO: failures here should probably not just end the renewal loop, there
+	// should be some retry / back-off logic.
+
+	// TODO: what should this interval be? should it be user configurable?
+	// Also, must be < the validity period.
+	// TODO: validate that cert is actually renewable.
+
+	log.Infof("Beginning renewal loop: ttl=%s interval=%s", cfg.CertificateTTL, cfg.RenewInterval)
+	if cfg.RenewInterval > cfg.CertificateTTL {
+		log.Errorf(
+			"Certificate TTL (%s) is shorter than the renewal interval (%s). The next renewal is likely to fail.",
+			cfg.CertificateTTL,
+			cfg.RenewInterval,
+		)
+	}
+
+	// Determine where the bot should write its internal data (renewable cert
+	// etc)
+	botDestination, err := cfg.Storage.GetDestination()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	ticker := time.NewTicker(cfg.RenewInterval)
+	defer ticker.Stop()
+	for {
+		log.Info("Attempting to renew bot certificates...")
+		newIdentity, err := renewIdentityViaAuth(client, ident, cfg.CertificateTTL)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		if err := describeIdentity(ident); err != nil {
+			return trace.WrapWithMessage(err, "Could not describe bot identity at %s", botDestination)
+		}
+
+		// TODO: warn if duration < certTTL? would indicate TTL > server's max renewable cert TTL
+		// TODO: error if duration < renewalInterval? next renewal attempt will fail
+
+		// Immediately attempt to reconnect using the new identity (still
+		// haven't persisted the known-good certs).
+		newClient, err := authenticatedUserClientFromIdentity(newIdentity, cfg.AuthServer)
```

Comment:
this calls `auth.NewClient` which dials in the background, you may not see an error until the first API call. Could try Ping or something if you want to make sure the identity is good